compiler/src/main/java/com/squareup/anvil/compiler/codegen/MergeAnnotationsCheckGenerator.kt

```diff
@@ -0,0 +1,76 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.google.auto.service.AutoService
+import com.squareup.anvil.compiler.api.AnvilContext
+import com.squareup.anvil.compiler.api.CodeGenerator
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference
+import com.squareup.anvil.compiler.internal.reference.AnvilCompilationExceptionClassReference
+import com.squareup.anvil.compiler.internal.reference.classAndInnerClassReferences
+import com.squareup.anvil.compiler.mergeComponentFqName
+import com.squareup.anvil.compiler.mergeInterfacesFqName
+import com.squareup.anvil.compiler.mergeModulesFqName
+import com.squareup.anvil.compiler.mergeSubcomponentFqName
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.psi.KtFile
+import java.io.File
+
+@AutoService(CodeGenerator::class)
+internal class MergeAnnotationsCheckGenerator : PrivateCodeGenerator() {
+  override fun isApplicable(context: AnvilContext): Boolean = !context.disableComponentMerging
+
+  override fun generateCodePrivate(
+    codeGenDir: File,
+    module: ModuleDescriptor,
+    projectFiles: Collection<KtFile>
+  ) {
+    projectFiles
+      .classAndInnerClassReferences(module)
+      .forEach { clazz ->
+        val annotations = clazz.annotations
+          .findAll(
+            mergeComponentFqName,
+            mergeSubcomponentFqName,
+            mergeModulesFqName,
+            mergeInterfacesFqName
+          )
+          .ifEmpty { return@forEach }
+
+        annotations.checkSingleAnnotation()
+        annotations.checkNoDuplicateScope(contributeAnnotation = false)
+
+        // Note that we only allow a single type of `@Merge*` annotation through the check above.
+        // The same class can't merge a component and subcomponent at the same time. Therefore,
+        // all annotations must have the same FqName and we can use the first annotation to check
+        // for the Dagger annotation.
+        annotations
+          .firstOrNull { it.fqName != mergeInterfacesFqName }
```

Comment:
Why do we check specifically for MergeInterfaces her? (compared to the other merge annotations that this generator applies to)

Reply:
Because this check below makes sure that you don't use `@MergeComponent` and `@dagger.Component` at the same time. Anvil will generate the Dagger component. There's no Dagger annotation for `@MergeInterfaces`, so this check doesn't make sense.